### PR TITLE
Add `Bittide.Counter`: a way of measuring domain speed differences

### PR DESCRIPTION
**Pull request by martijnbastiaan**
Exposes a single function, `domainDiffCounter`, that we can use to measure speed differences between two domains. These can in turn be used to eliminate the need for using FIFOs directly to measure clock divergence.

TODO: 

 - [x] Tests
 - [x] Instance?

### CODE REVIEW

File: bittide/src/Bittide/Counter.hs [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+import Clash.Cores.Xilinx.Xpm (xpmCdcGray)
+import Clash.Sized.Extra (unsignedToSigned, glueUnsigneds)
+
+data DdcState
+  = DdcInReset
+  | DdcRunning (Unsigned 64)
+  deriving (Generic, NFDataX)
+
+-- | Indicates whether 'domainDiffCounter' is actively counting or still in reset.
+type Active = Bool
```
Comment (kleinreact) [inline bittide/src/Bittide/Counter.hs:21]: `Active` is not used currently.

File: bittide/src/Bittide/Counter.hs [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+  , domainDiffCounter
+  ) where
+
+import Clash.Explicit.Prelude
+
+import Clash.Cores.Xilinx.Xpm (xpmCdcGray)
+import Clash.Sized.Extra (unsignedToSigned, glueUnsigneds)
+
+-- | State of 'domainDiffCounter'
+data DdcState
```
Comment (kleinreact) [inline bittide/src/Bittide/Counter.hs:16]: This definition matches `EscState`, so maybe reduce it to a single one. This makes it also easier to notice the 64 bit size constraint, while this way a reader alwasy has to keep in mind that `DdcState` is hardcoded to be 64 bits.

Comment (kleinreact) [reply bittide/src/Bittide/Counter.hs:16]: Other than that, its basically just `Maybe`. Decide on your own whether you like to keep use of that fact.

File: bittide/src/Bittide/Counter.hs [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+-- __N.B.__: This function uses an 8-bit Gray counter internally, and will therefore
+--           only work properly if both clock speeds are pretty close to one another.
+synchronizedSuccCounter ::
+  forall n m src dst .
+  ( KnownDomain src
+  , KnownDomain dst
+  , KnownNat n
+  , 8 <= n
+
+  -- TODO: Haxiom this out:
```
Comment (kleinreact) [inline bittide/src/Bittide/Counter.hs:88]: Don't forget.

File: bittide/tests/Tests/Counter.hs [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+      sampledNonZero = P.dropWhile (==0) sampled
+      len = P.length sampledNonZero
+    assertBool (">1 @ " <> show n) (len > 1)
+    assertEqual ("exp @ " <> show n) sampledNonZero (P.take len expected)
+ where
+  dut = top @D10 @D20 noRst
+  expected = P.concat [[n, n] | n <- [1..]]
+
+tests :: TestTree
+tests = $(testGroupGenerator)
```
Comment (kleinreact) [inline bittide/tests/Tests/Counter.hs:71]: I its not too much of an effort, also testing the overflow of the `Signed 16` output and the internal `Unsigned 64` counter would be great. It probably requires making these numbers smaller for testing. It depends on the open implementation efforts regarding my previous comments.

**Review by kleinreact — CHANGES_REQUESTED** [review]
I am a bit worried about the overflow of the internal counters. Is this not handled on purpose, or have I missed something such that this is no issue in our case?

File: bittide/src/Bittide/Counter.hs [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+  counter = synchronizedSuccCounter @64 clkSrc rstSrc clkDst rstDst
+
+  go :: DdcState -> Unsigned 64 -> (DdcState, (Signed 16, Bool))
+  go DdcInReset c1
+    | c1 == 0   = (DdcInReset,          (0, False))
+    | otherwise = (DdcRunning (c1 + 1), (0, True))
+  go (DdcRunning c0) c1 = (DdcRunning (c0 + 1), (c0 `subAndTruncate` c1, True))
+
+  subAndTruncate :: Unsigned 64 -> Unsigned 64 -> Signed 16
+  subAndTruncate c0 c1 = truncateB (unsignedToSigned c0 - unsignedToSigned c1)
```
[the remark this answers is not in the document]
Comment (martijnbastiaan) [reply bittide/src/Bittide/Counter.hs:73]: You're right, I didn't do my back-of-a-napkin calculation right. The goal of this would to basically never overflow in a Bittide system. Given our clock setup we know that the maximum divergence is +-100 ppm. On a 200 MHz clock this amounts to 40k ticks difference at most, every second. A 16 bit number would therefore quickly overflow (~3 seconds).

I think it's therefore best to up this to a `Signed 32`. This would allow a system to run a maximum divergence rate for 2.5 days. This should be more than enough given that neither the rate nor that duration is realistic - clock control should have kicked in way before that.

File: bittide/src/Bittide/Counter.hs (hunk 3 above)
Comment (martijnbastiaan) [reply bittide/src/Bittide/Counter.hs:88]: Apparently this was only necessary in earlier iterations of the code :shrug:

File: bittide/src/Bittide/Counter.hs (hunk 5 above)
[the remark this answers is not in the document]
Comment (martijnbastiaan) [reply bittide/src/Bittide/Counter.hs:84]: We decided that this is enough for now.

**Review by kleinreact — APPROVED** [review]
LGTM then.
